### PR TITLE
Add podManifestTemplate as devpod customizations in devcontainer.json

### PR DESCRIPTION
**Pull request by titilambert**
This PR should fix:
* https://github.com/loft-sh/devpod/issues/546
* https://github.com/loft-sh/devpod/issues/534
* https://github.com/loft-sh/devpod/issues/401
* https://github.com/loft-sh/devpod/issues/390
* https://github.com/loft-sh/devpod/issues/362

Here is an example of a `devcontainer.json`
```json
{
  "name": "my project",
  "image": "ubuntu:focal-20230624",
  "customizations": {
      "devpod": {
             "podManifestTemplate": ".devcontainer/pod_manifest.yaml"
          }
}
```

Here is an example of the `devcontainer/pod_manifest.yaml`
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: USELESS
  namespace: USELESS
  labels:
    mylabel: myvalue
spec:
  tolerations:
    - effect: ...
      key: ...
      operator: ...
  containers:
  - args:
    - USELESS
    command:
    - USELESS
    image: USELESS
    name: USELESS
    resources:
      limits:
        cpu: 200m
      requests:
        cpu: "200m"
    volumeMounts:
    - mountPath: /data
      name: data
      readOnly: true
  volumes:
    - name: data
      hostPath:
        path: /data
        type: Directory
```

In this example, I'm adding labels/volumes/volumeMounts/resources/tolerations but we could add more than that since the new code will merge the pod manifest template with the needed settings to get the pod working as needed (like the name, args, command, ...)
NOTE: there is some useless settings like `image` since it will be replaced by the one set by the code itself

### CODE REVIEW

**Comment by 89luca89** [comment]
Hi @titilambert thanks a lot for the PR!

There are some issues, I think it's missing the new dependencies in the vendor directory

**Comment by pascalbreuninger** [comment]
Hey @titilambert, thanks for the PR! The solution looks good to me (although tests are failing, I think you'll need to push the new vendored dependencies).

Do you have the strict use-case of this being on the workspace level? 
Imo this should be merged into the kubernetes provider rather than the `devcontainer.json` to keep the building of the devcontainer clearly separated from the infrastructure it is running on

**Comment by titilambert** [comment]
@pascalbreuninger /

> Do you have the strict use-case of this being on the workspace level?

@pascalbreuninger not really, I was not sure where to put the settings since some of them like the image are in the `devcontainer.json` file and some others in the providers.
I did put it in the `devcontainer.json` file just because I was thinking that the pod manifest file could be stored in a different folder on each git repository, so I thought it was more explicit fo the devs if the setting is in the `devcontainer.json` file.

@pascalbreuninger what do you think ?

**Comment by pascalbreuninger** [comment]
Imo this could be a kubernetes provider option, similar to [kubeconfig](https://github.com/loft-sh/devpod/blob/95b87e384a4e4f447666caa89e427b063c50812b/providers/kubernetes/provider.yaml#L53C3-L53C20), so you'd reference a file on your local machine when setting up the provider

**Comment by titilambert** [comment]
> Imo this could be a kubernetes provider option, similar to [kubeconfig](https://github.com/loft-sh/devpod/blob/95b87e384a4e4f447666caa89e427b063c50812b/providers/kubernetes/provider.yaml#L53C3-L53C20), so you'd reference a file on your local machine when setting up the provider

The kubeconfig file is not related to a project (a git repo) and it's a personal file (since we could have creds inside).
The pod manifest seems to be related to a project (at least the file path), since it will be different, depending on the project (git repo)

**Comment by titilambert** [comment]
> Hi @titilambert thanks a lot for the PR!
> 
> There are some issues, I think it's missing the new dependencies in the vendor directory

Sorry ! Should be fixed now.

**Comment by pascalbreuninger** [comment]
> > Imo this could be a kubernetes provider option, similar to [kubeconfig](https://github.com/loft-sh/devpod/blob/95b87e384a4e4f447666caa89e427b063c50812b/providers/kubernetes/provider.yaml#L53C3-L53C20), so you'd reference a file on your local machine when setting up the provider
> 
> The kubeconfig file is not related to a project (a git repo) and it's a personal file (since we could have creds inside). The pod manifest seems to be related to a project (at least the file path), since it will be different, depending on the project (git repo)

I don't agree, from my point of view both the pod manifest and the kubeconfig are concerned with how you create the infrastructure for your devcontainer whereas the configuration files within `/devcontainer` determines how you build the devcontainer itself.

Imagine a scenario where you'd have nodeSelectors on your pod manifest but then someone wants to run your project on another cluster without the expected nodes, it'll break for them. 
If we keep this on the provider level however, you can still customize the pod its running on and potentially share that provider with others, without requiring it though.

**Comment by titilambert** [comment]
@pascalbreuninger I agree with you !
I will move it as provider setting today.
I just have one question left: If the path set in the provider settings is not absolute, should I still use the git repo local folder path as base path ?

**Comment by pascalbreuninger** [comment]
phew, good question :D
based on the problems we had with `kubeconfig` and the path there I'd always treat it as absolute if possible 
(see [this](https://loft-sh.slack.com/archives/C056ZDZPJ4W/p1687352153935179) slack thread for context).
You could add a check to the providers `init` command to make sure we're able to find the pod manifest if specified

**Comment by titilambert** [comment]
@pascalbreuninger I changed it to load it from the provider settings.
I'm not sure about what is `provider init` command.
But let me know if the code change is good enough.

Thanks

**Comment by pascalbreuninger** [comment]
@titilambert only had 2 smaller comments, other than that it looks good to me. Once the comments are addressed we can merge :) Thank you!

File: pkg/driver/kubernetes/run.go [hunk 1]
```diff
@@ -170,38 +170,49 @@ func (k *kubernetesDriver) runContainer(
...
+			return err
+		}
+	}
+
+	var pod *corev1.Pod
+	if podTemplate != nil {
+		pod = podTemplate
+		pod.ObjectMeta.Name = id
+		pod.ObjectMeta.Namespace = ""
+		pod.ObjectMeta.Labels = map[string]string{}
```
Comment (neogopher) [inline pkg/driver/kubernetes/run.go:188]: We are resetting the pod Labels here, but I did not find any place where we are setting them back from the podTemplate and/or extraLabels from config. In such case, we may loose all labels or am I missing something here.

**Review by neogopher — COMMENTED** [review]
Hi @titilambert thanks a lot for this PR!! Loved your idea. :raised_hands: 
Just my 2 cents below

**Comment by titilambert** [comment]
> @titilambert only had 2 smaller comments, other than that it looks good to me. Once the comments are addressed we can merge :) Thank you!

I don't see your comments. did you forgot to validate them ?

**Review by pascalbreuninger — CHANGES_REQUESTED** [review]

**Comment by pascalbreuninger** [comment]
@titilambert wow yes I did, thank you!

File: pkg/driver/kubernetes/run.go [hunk 2]
```diff
@@ -170,38 +170,49 @@ func (k *kubernetesDriver) runContainer(
...
+			}
+		}
+		pod.Spec.ServiceAccountName = serviceAccount
+
+		pod.Spec.InitContainers = append(initContainer, podTemplate.Spec.InitContainers...)
+		pod.Spec.Containers[0].Name = "devpod"
+		pod.Spec.Containers[0].Image = imageName
+		pod.Spec.Containers[0].Command = []string{entrypoint}
+		pod.Spec.Containers[0].Args = args
+		pod.Spec.Containers[0].Env = append(envVars, podTemplate.Spec.Containers[0].Env...)
```
[the remark this answers is not in the document]
Comment (titilambert) [reply pkg/driver/kubernetes/run.go:204]: I don't get what you mean here

[the remark this answers is not in the document]
Comment (pascalbreuninger) [reply pkg/driver/kubernetes/run.go:204]: We set the `spec.containers[0].resources` field in the `else` branch but not in here, is this on purpose?

[the remark this answers is not in the document]
Comment (titilambert) [reply pkg/driver/kubernetes/run.go:204]: I forget it ... Thanks !

File: pkg/driver/kubernetes/run.go (hunk 1 above)
Comment (titilambert) [reply pkg/driver/kubernetes/run.go]: fixed

File: pkg/driver/kubernetes/helper.go [hunk 3]
```diff
@@ -56,6 +58,24 @@ func parseResources(resourceString string, log log.Logger) corev1.ResourceRequir
 	}
 }
 
+func getPodTemplate(manifestFilePath string) (*corev1.Pod, error) {
+	body, err := ioutil.ReadFile(manifestFilePath)
```
[the remark this answers is not in the document]
Comment (titilambert) [reply pkg/driver/kubernetes/helper.go]: fixed

File: pkg/driver/kubernetes/run.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (titilambert) [reply pkg/driver/kubernetes/run.go:200]: fixed

[the remark this answers is not in the document]
Comment (pascalbreuninger) [reply pkg/driver/kubernetes/run.go:200]: not sure if I'm blind or if you forgot to push but I can't see the `resources` 😅

**Review by pascalbreuninger — CHANGES_REQUESTED** [review]

**Comment by pascalbreuninger** [comment]
@titilambert the labels need yet another change, thanks for the patience 🙇

File: pkg/driver/kubernetes/run.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (titilambert) [reply pkg/driver/kubernetes/run.go:200]: It's there:
https://github.com/loft-sh/devpod/pull/547/files#diff-6d6151bffaea4d9680c3662e730e4cfb80d97a5d6d69a0102a7a33fa0f23c027R209-R211

**Comment by titilambert** [comment]
> @titilambert the labels need yet another change, thanks for the patience bow

No problem ! and thanks you for your patience too !

**Review by pascalbreuninger — APPROVED** [review]
LGTM

**Pull request merged**